src/Tinc/PackageDb.hs

```diff
@@ -0,0 +1,50 @@
+{-# LANGUAGE CPP #-}
+module Tinc.PackageDb (
+  PackageDb(..)
+, PackageConfig
+, readPackageDb
+, lookupPackageConfig
+, allPackageConfigs
+#ifdef TEST
+, packageFromPackageConfig
+#endif
+) where
+
+import           Prelude ()
+import           Prelude.Compat
+
+import           Control.Exception
+import           Data.List.Compat
+import           Data.Map (Map)
+import qualified Data.Map as Map
+import           System.Directory
+import           System.FilePath
+
+import           Package
+import           Tinc.Types
+
+data PackageConfig
+
+data PackageDb = PackageDb {
+  _packageDbPath :: Path PackageDb
+, packageDbPackageConfigs :: Map Package (Path PackageConfig)
+} deriving (Eq, Show)
+
+readPackageDb :: Path PackageDb -> IO PackageDb
+readPackageDb p = do
+  packageConfigs <- filter (".conf" `isSuffixOf`) <$> getDirectoryContents (path p)
+  let packages = map packageFromPackageConfig packageConfigs
+  return $ PackageDb p (Map.fromList $ zip packages (map (Path . (path p </>)) packageConfigs))
```

Comment:
I think I would try to use one map that produces `(k, v)`. So that you end up with `return $ PackageDb p (Map.fromList $ map _ packageConfigs)` (with the hole filled in).